Data/HashMap/Internal.hs

```diff
@@ -1679,6 +1682,28 @@ unions :: (Eq k, Hashable k) => [HashMap k v] -> HashMap k v
 unions = L.foldl' union empty
 {-# INLINE unions #-}
 
+
+------------------------------------------------------------------------
+-- * Compose
+
+-- | Relate the keys of one map to the values of
+-- the other, by using the values of the former as keys for lookups
+-- in the latter.
+--
+-- Complexity: \( O (n * \log(m)) \), where \(m\) is the size of the first argument
+--
+-- > compose (fromList [('a', "A"), ('b', "B")]) (fromList [(1,'a'),(2,'b'),(3,'z')]) = fromList [(1,"A"),(2,"B")]
+--
+-- @
+-- ('compose' bc ab '!?') = (bc '!?') <=< (ab '!?')
+-- @
+--
+-- @since UNRELEASED
+compose :: (Eq b, Hashable b) => HashMap b c -> HashMap a b -> HashMap a c
+compose bc !ab
+  | null bc = empty
+  | otherwise = mapMaybe (bc !?) ab
```

Comment:
Since we have both strict and lazy variants of `mapMaybe`, shouldn't `compose` have both variants too?

Reply:
Hmmm.... My intuition suggests no. What did we do for `Data.Map`?

Reply:
For `Map` and `IntMap`, we have both variants.

Reply:
:-/.

Reply:
It's kind of weird, but I guess it makes sense to be able to fuse "compose two maps" with "force all the values in the result". I definitely think the surprising strictness should be documented explicitly.

Reply:
I can't quite tell what the decision was - do I add the other variant or ?

Reply:
Yeah. I didn't understand @treeowl's comment above.

@treeowl Could you clarify what your preferred solution is, and possibly why we shouldn't have two variants as in `containers`?

Reply:
I think we probably *should* add it for consistency without user-facing semantic churn. And it is a *potentially* useful function. But it's really quite *weird*, and it's likely *not* what you actually want most of the time. Suppose you have maps `m` and `n`, and `m` was build using the `Strict` API, so all the values are in WHNF. When you `compose m n` using the *strict* `compose`, then it will unnecessarily force all the relevant values in `m` again. Thanks to pointer tagging, that probably isn't actually going to cost much, but it seems a tad silly. Conclusion: add the strict version, but note in its Haddock documentation that the *lazy* version is generally preferable if the *first* map's values are all evaluated.

Reply:
Thanks for the explanation, @treeowl! :)

I had initially believed that a version using the lazy `mapMaybe` might prevent the first argument from being gc'd for longer than necessary. I now realized that we need to distinguish `Just`s from `Nothing`s to build the spine too.

It now seems to me that a version using the strict `mapMaybe` would indeed only add value when the first map has unevaluated values.

I'm curious now: Are there other functions in the `*.Strict` modules that operate under the assumption that the values of any input maps are already in WHNF?

Reply:
These are slippery questions! Arguably `union` and such.

Reply:
After thinking about this some more, I think it makes more sense to avoid re-evaluation, like in `union` etc.

`containers` should be fixed to avoid re-evaluating the values too.

Or maybe I don't understand the original motivation for having strict variants in `containers`. @alexfmpe, can you comment on this?

Reply:
I don't quite recall since it was over a year ago, but I suspect it wasn't particularly thought through

Reply:
@treeowl How do you feel about aligning `compose` with `union` etc. and removing its strict variants from `containers`?

Reply:
Yeah, I think that probably makes sense.

Reply:
I have prepared a fix for `containers` in https://github.com/haskell/containers/pull/745.

Reply:
@alexfmpe Could you update your PR to remove the strict variants again?